adapter/pkg/synchronizer/client_pool.go

```diff
@@ -0,0 +1,188 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Package "synchronizer" contains artifacts relate to fetching APIs and
+ * API related updates from the control plane event-hub.
+ * This file contains functions to retrieve APIs and API updates.
+ */
+
+package synchronizer
+
+import (
+	"crypto/tls"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/wso2/product-microgateway/adapter/pkg/loggers"
+	"github.com/wso2/product-microgateway/adapter/pkg/tlsutils"
+)
+
+type worker struct {
+	id              int
+	internalQueue   <-chan workerRequest
+	processFunc     processHTTPRequest
+	delayAfterFault time.Duration
+}
+
+// workerRequest is the task which can be submitted to the pool.
+type workerRequest struct {
+	Req                http.Request
+	APIUUID            *string
+	labels             []string
+	SyncAPIRespChannel chan SyncAPIResponse
+}
+
+// pool is the worker pool which is handling
+type pool struct {
+	// TODO: (VirajSalaka) remove timeout
+	internalQueue      chan workerRequest
+	workers            []*worker
+	quit               chan bool
+	timeout            time.Duration
+	client             http.Client
+	controlPlaneParams controlPlaneParameters
+}
+
+type controlPlaneParameters struct {
+	serviceURL    string
+	username      string
+	password      string
+	retryInterval time.Duration
+}
+
+type processHTTPRequest func(*http.Request, *string, []string, chan SyncAPIResponse, *http.Client) bool
+
+func (w *worker) ProcessFunction() {
+	for workerReq := range w.internalQueue {
+		responseReceived := w.processFunc(&workerReq.Req, workerReq.APIUUID, workerReq.labels, workerReq.SyncAPIRespChannel,
+			&workerPool.client)
+		if !responseReceived {
+			time.Sleep(w.delayAfterFault)
+		}
+	}
+}
+
+var (
+	// WorkerPool is the thread pool responsible for sending the control plane request to fetch APIs
+	workerPool        *pool
+	oncePoolInitiated sync.Once
+)
+
+// InitializeWorkerPool creates the Worker Pool used for the Control Plane Rest API invocations.
+// maxWorkers indicate the maximum number of parallel workers sending requests to the control plane.
+// jobQueueCapacity indicate the maximum number of requests can kept inside a single worker's queue.
+// delayForFaultRequests indicate the delay a worker enforce (in seconds) when a fault response is received.
+func InitializeWorkerPool(maxWorkers, jobQueueCapacity int, delayForFaultRequests time.Duration, trustStoreLocation string,
+	skipSSL bool, requestTimeout, retryInterval time.Duration, serviceURL, username, password string) {
```

Comment:
retryInterval usage?

Reply:
It is used for the retry interval engaged for the fault requests which needs to be retried.